hbase-server/src/main/java/org/apache/hadoop/hbase/master/AssignmentManager.java

```diff
@@ -1972,6 +1975,13 @@ private void unassign(final HRegionInfo region,
       final RegionState state, final int versionOfClosingNode,
       final ServerName dest, final boolean transitionInZK,
       final ServerName src) {
+    String encodedName = region.getEncodedName();
+    AtomicInteger failedCloseCount = failedCloseTracker.get(encodedName);
+    if (failedCloseCount == null) {
+      failedCloseCount = new AtomicInteger();
+      failedCloseTracker.put(encodedName, failedCloseCount);
```

Comment:
We can race between test and put, potentially undercounting (by ref overwrite). Probably should test if the return of putIfAbsent is not null, and use/update that. If the code that updates `failedOpenTracker` also has this racy pattern it should be fixed too.

Reply:
I agree, that's what we do even for failedOpenTracker. I can make the change for forceRegionStateToOffline() to take the lock before changing the state

Reply:
That's fine